pkg/ccl/backupccl/backup_processor.go

```diff
@@ -188,277 +197,409 @@ func runBackupProcessor(
+		data, err = storageccl.EncryptFile(data, s.conf.enc.Key)
 		if err != nil {
 			return err
 		}
 	}
 
-	if err := cloud.WriteFile(ctx, exportStore, file.Path, bytes.NewReader(data)); err != nil {
+	name := fmt.Sprintf("%d.sst", builtins.GenerateUniqueInt(s.conf.id))
+	if err := cloud.WriteFile(ctx, s.dest, name, bytes.NewReader(data)); err != nil {
 		log.VEventf(ctx, 1, "failed to put file: %+v", err)
 		return errors.Wrap(err, "writing SST")
 	}
+
+	for i := range s.bufferedFiles {
+		s.bufferedFiles[i].Path = name
+	}
+
+	progDetails := BackupManifest_Progress{
+		RevStartTime: s.bufferedRevStart,
+		Files:        s.bufferedFiles,
+		Partial:      !s.finishedSpan,
+	}
+	var prog execinfrapb.RemoteProducerMetadata_BulkProcessorProgress
+	details, err := gogotypes.MarshalAny(&progDetails)
+	if err != nil {
+		return err
+	}
+	prog.ProgressDetails = *details
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case s.conf.progCh <- prog:
+	}
+
+	s.sstWriter = nil
+	s.bufferedSST.Reset()
+	s.bufferedFiles = nil
+	s.bufferedRevStart.Reset()
+	s.finishedSpan = false
+
+	return nil
+}
+
+func (s *sstSink) write(ctx context.Context, resp returnedSST) error {
+	span := resp.f.Span
+
+	// If this span starts before the last buffered span ended, we need to flush
+	// since it overlaps but SSTWriter demands writes in-order.
+	if len(s.bufferedFiles) > 0 && span.Key.Compare(s.bufferedFiles[len(s.bufferedFiles)-1].Span.EndKey) < 0 {
+		if err := s.flush(ctx); err != nil {
+			return err
+		}
+	}
+
+	// Initialize the writer if needed then copy the SST content to the writer.
+	if s.sstWriter == nil {
+		w := storage.MakeBackupSSTWriter(&s.bufferedSST)
+		s.sstWriter = &w
+	}
+	sst, err := storage.NewMemSSTIterator(resp.sst, false)
+	if err != nil {
+		return err
+	}
+	defer sst.Close()
+	sst.SeekGE(storage.MVCCKey{Key: keys.MinKey})
+	for {
+		if valid, err := sst.Valid(); !valid || err != nil {
+			if err != nil {
+				return err
+			}
+			break
+		}
+		k := sst.UnsafeKey()
+		if k.Timestamp.IsEmpty() {
+			if err := s.sstWriter.PutUnversioned(k.Key, sst.UnsafeValue()); err != nil {
+				return err
+			}
+		} else {
+			if err := s.sstWriter.PutMVCC(sst.UnsafeKey(), sst.UnsafeValue()); err != nil {
+				return err
+			}
+		}
+		sst.Next()
+	}
+
+	// If this span extended the last span added -- that is, picked up where it
+	// ended and has the same time-bounds -- then we can simply extend that span
+	// and add to its entry counts. Otherwise we need to record it separately.
+	if l := len(s.bufferedFiles) - 1; l > 0 && s.bufferedFiles[l].Span.EndKey.Equal(span.Key) &&
+		s.bufferedFiles[l].EndTime.EqOrdering(resp.f.EndTime) &&
+		s.bufferedFiles[l].StartTime.EqOrdering(resp.f.StartTime) {
+		s.bufferedFiles[l].Span.EndKey = span.EndKey
+		s.bufferedFiles[l].EntryCounts.add(resp.f.EntryCounts)
+	} else {
+		s.bufferedFiles = append(s.bufferedFiles, resp.f)
+	}
+	s.bufferedRevStart.Forward(resp.revStart)
+	s.finishedSpan = s.finishedSpan || resp.finished
```

Comment:
I think I convinced myself that progress updates only need to be marked as partial if all buffered `resp`s are not `finished`, but is it okay for the progress updater to only get 1 `progDetails` if multiple `finished` spans are batched together? Perhaps a comment on `finishedSpan`'s declaration on `sstSink`'s type declaration would clarify what `finishedSpan` means in terms of the buffered files.

Reply:
Reworked a little in light of #66654 so I think it is clearer and more correct now: we track how many spans a progress update completes, which for a normal one-file-finished-one-span update, is just 1 and is 0 for resumeSpans!=nil spans, but has the nice property of being able to be summed for our buffer as well if it contains, say, 10 span-finishing files.